aptos-move/framework/aptos-framework/sources/storage_gas.move

```diff
@@ -0,0 +1,359 @@
+    ///
+    /// Specifically, it is updated by executing a reconfig transaction based
+    /// on the usage at the begining of the current epoch. The gas schedule
+    /// derived from these parameter will be for gas calculation of the entire
+    /// next epoch.
+    /// -- The data is one epoch older than ideal, but VM doesn't need to worry
+    /// about reloading gas parameters after the first txn of an epoch.
+    struct StorageGas has key {
+        per_item_read: u64,
+        per_item_create: u64,
+        per_item_write: u64,
+        per_byte_read: u64,
+        per_byte_create: u64,
+        per_byte_write: u64,
+    }
+
+    // x and y are basis points.
+    struct Point has copy, drop, store {
+        x: u64,
+        y: u64
+    }
+
+    struct UsageGasConfig has copy, drop, store {
+        target_usage: u64,
+        read_curve: GasCurve,
+        create_curve: GasCurve,
+        write_curve: GasCurve,
+    }
+
+    /// The curve assumes there are two points (0, 0) and (10000, 10000) on both ends. Moreover, points must also
+    /// satisfy the following rules:
+    /// 1. the x values must be strictly increasing and between (0, 10000);
+    /// 2. the y values must be non-decreasing and between (0, 10000);
+    /// So the curve will be identified as point (0, 0) and (10000, 10000) interpolated with the points. The y value
+    /// between two points will be calculated by neighboring points as if there is a linear line connecting these two
+    /// points.
+    struct GasCurve has copy, drop, store {
+        min_gas: u64,
+        max_gas: u64,
+        points: vector<Point>,
+    }
+
+    /// P(x) = min_price + (base ^ (utilization / target_usage) - 1) / (base - 1) * (max_price - min_price)
+    // Provide a default exponential curve with the base to be 8192, which means:
+    // When DB is at 50% target usage,the price increases roughly 1.09% of (max_price - min_price).
+    // Detailed data points:
+    // 10% -> 0.02%
+    // 20% -> 0.06%
+    // 30% -> 0.17%
+    // 40% -> 0.44%
+    // 50% -> 1.09%
+    // 60% -> 2.71%
+    // 70% -> 6.69%
+    // 80% -> 16.48%
+    // 90% -> 40.61%
+    // 95% -> 63.72%
+    // 99% -> 91.38%
+    public fun base_8192_exponential_curve(min_gas: u64, max_gas: u64): GasCurve {
+        new_gas_curve(min_gas, max_gas,
+            vector[
+                new_point(1000, 2),
+                new_point(2000, 6),
+                new_point(3000, 17),
+                new_point(4000, 44),
+                new_point(5000, 109),
+                new_point(6000, 271),
+                new_point(7000, 669),
+                new_point(8000, 1648),
+                new_point(9000, 4061),
+                new_point(9500, 6372),
+                new_point(9900, 9138),
+            ]
+        )
+    }
+
+    struct StorageGasConfig has copy, drop, key {
+        item_config: UsageGasConfig,
+        byte_config: UsageGasConfig,
+    }
+
+    public fun new_point(x: u64, y: u64): Point {
```

Comment:
Does this need to be public? Same question for new_gas_curve and new_usage_gas_config

Reply:
if it is not public there is no way to construct a final `new_storage_gas_config` and pass it to `set_config` function from a on-chain governance txn, right?

Reply:
Oh I see. Fine then